daemon/endpoint.go

```diff
@@ -211,6 +211,44 @@ func (h *putEndpointID) Handle(params PutEndpointIDParams) middleware.Responder
 	if err != nil {
 		apierror.Error(code, err)
 	}
+
+	// Wait for endpoint to be ready if specified in API call.
+	if params.Endpoint.SyncBuildEndpoint {
+		log.Debug("synchronously waiting for endpoint to be in %s state", endpoint.StateReady)
+
+		e, err := endpointmanager.Lookup(params.ID)
+		if err != nil {
+			return apierror.Error(PutEndpointIDFailedCode, err)
+		}
+
+		// Default timeout for PUT /endpoint/{id} is 30 seconds, so put  timeout
+		// in this function a bit below that timeout. If the timeout for clients
+		// in API is below this value, they will get a message containing
+		// "context deadline exceeded".
+		stateChangeTimeout := time.Duration(25 * time.Second)
+
+		// Check up until stateChangeTimeout seconds for endpoint state before
+		// waiting for endpoint to be in ready state.
+		timeout := time.After(stateChangeTimeout)
+
+		// Check for endpoint state every second.
+		ticker := time.NewTicker(1 * time.Second)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				e.Mutex.Lock()
+				epState := e.GetStateLocked()
+				e.Mutex.Unlock()
+				if epState == endpoint.StateReady {
+					return nil
+				}
+			case <-timeout:
+				return apierror.Error(PutEndpointIDFailedCode, fmt.Errorf("endpoint did not synchronously regenerate after timeout"))
```

Comment:
Same here, endpoint must be removed again

Reply:
Done.